libmscore/scale.cpp

```diff
@@ -0,0 +1,265 @@
+//   loadScale
+//---------------------------------------------------------
+
+bool Scale::loadScale(const QString& fn)
+      {
+      static set<int> supportedNbs = { 8, 12, 21, 35 };
+
+      QFile f(fn);
+      if (!f.open(QIODevice::ReadOnly)) {
+            MScore::lastError = strerror(errno);
+            return false;
+            }
+
+      QTextStream in(&f);
+
+      // Read description of the scala file
+      QString description = getNextLine(in);
+
+      int nbNotes = getNextLine(in).toInt();
+      if (supportedNbs.find(nbNotes + 1) == supportedNbs.end()) {
+            MScore::lastError = QObject::tr("MuseScore supports only Scala files with 8, 12, 21 or 35 notes");
+            return false;
+            }
+
+      for (int i = 0; i < nbNotes; ++i) {
+            QString pitchLine = getNextLine(in);
+            if (pitchLine.contains('/')) {
+                  QStringList values = pitchLine.split('/');
+                  if (values.size() != 2 || values[1].toDouble() == 0) {
+                        MScore::lastError = QObject::tr("Unsupported Scala file format.");
+                        return false;
+                        }
+
+                  originalNotes.push_back(1200 * log(values[0].toDouble() / values[1].toDouble()) / log(2));
+            } else {
+                  originalNotes.push_back(pitchLine.toDouble());
+                  }
+            }
+
+      computedNotes.clear();
+      switch (originalNotes.size()) {
+            case 7:
+                  computeNotes7();
+                  break;
+            case 12:
+                  computeNotes12();
+                  break;
+            case 17:
+                  computeNotes17();
+                  break;
+            case 21:
+                  computeNotes21();
+                  break;
+            case 35:
+                  computeNotes35();
+                  break;
+            default:
+                  MScore::lastError = QObject::tr("Unsupported Scala file format.");
+                  return false;
+            }
+      return true;
+      }
+
+//---------------------------------------------------------
+//   computeNotes8
+//---------------------------------------------------------
+void Scale::computeNotes7() {
+      static vector<Tpc> notesToFill = {TPC_D, TPC_E, TPC_F, TPC_G, TPC_A, TPC_B, TPC_C};
+
+      for (size_t i = 0; i < originalNotes.size(); ++i) {
+            computedNotes[notesToFill[i]] = originalNotes[i];
+            }
+
+      fillMissing7();
+      }
+
+//---------------------------------------------------------
+//   computeNotes12
+//---------------------------------------------------------
+void Scale::computeNotes12() {
+      static vector<Tpc> notesToFill = {
+            TPC_C_S, TPC_D,
+            TPC_D_S, TPC_E,
+                     TPC_F,
+            TPC_F_S, TPC_G,
+            TPC_G_S, TPC_A,
+            TPC_A_S, TPC_B,
+                     TPC_C};
+
```

Comment:
The most common attribution of black keys is: C#, Eb, F#, G# (occasionally Ab), Bb